# **Homework**: Data talks club data engineering zoomcamp Data loading workshop

Hello folks, let's practice what we learned - Loading data with the best practices of data engineering.

Here are the exercises we will do




# 1. Use a generator

Remember the concept of generator? Let's practice using them to futher our understanding of how they work.

Let's define a generator and then run it as practice.

**Answer the following questions:**

- **Question 1: What is the sum of the outputs of the generator for limit = 5?**
- **Question 2: What is the 13th number yielded**

I suggest practicing these questions without GPT as the purpose is to further your learning.

In [49]:
def square_root_generator(limit):
    n = 1
    while n <= limit:
        yield n ** 0.5
        n += 1

# Example usage:
limit = 5
generator = square_root_generator(limit)

for sqrt_value in generator:
    print(sqrt_value)


1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
8.382332347441762


# 2. Append a generator to a table with existing data


Below you have 2 generators. You will be tasked to load them to duckdb and answer some questions from the data

1. Load the first generator and calculate the sum of ages of all people. Make sure to only load it once.
2. Append the second generator to the same table as the first.
3. **After correctly appending the data, calculate the sum of all ages of people.**




In [ ]:
def people_1():
    for i in range(1, 6):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 25 + i, "City": "City_A"}

for person in people_1():
    print(person)


def people_2():
    for i in range(3, 9):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 30 + i, "City": "City_B", "Occupation": f"Job_{i}"}


for person in people_2():
    print(person)


{'ID': 1, 'Name': 'Person_1', 'Age': 26, 'City': 'City_A'}
{'ID': 2, 'Name': 'Person_2', 'Age': 27, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 28, 'City': 'City_A'}
{'ID': 4, 'Name': 'Person_4', 'Age': 29, 'City': 'City_A'}
{'ID': 5, 'Name': 'Person_5', 'Age': 30, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 33, 'City': 'City_B', 'Occupation': 'Job_3'}
{'ID': 4, 'Name': 'Person_4', 'Age': 34, 'City': 'City_B', 'Occupation': 'Job_4'}
{'ID': 5, 'Name': 'Person_5', 'Age': 35, 'City': 'City_B', 'Occupation': 'Job_5'}
{'ID': 6, 'Name': 'Person_6', 'Age': 36, 'City': 'City_B', 'Occupation': 'Job_6'}
{'ID': 7, 'Name': 'Person_7', 'Age': 37, 'City': 'City_B', 'Occupation': 'Job_7'}
{'ID': 8, 'Name': 'Person_8', 'Age': 38, 'City': 'City_B', 'Occupation': 'Job_8'}


# 3. Merge a generator

Re-use the generators from Exercise 2.

A table's primary key needs to be created from the start, so load your data to a new table with primary key ID.

Load your first generator first, and then load the second one with merge. Since they have overlapping IDs, some of the records from the first load should be replaced by the ones from the second load.

After loading, you should have a total of 8 records, and ID 3 should have age 33.

Question: **Calculate the sum of ages of all the people loaded as described above.**


# Solution: First make sure that the following modules are installed:

In [ ]:
#Install the dependencies
%%capture
!pip install dlt[duckdb]

In [ ]:
# to do: homework :)

Questions? difficulties? We are here to help.
- DTC data engineering course channel: https://datatalks-club.slack.com/archives/C01FABYF2RG
- dlt's DTC cohort channel: https://dlthub-community.slack.com/archives/C06GAEX2VNX

---

# Solutions

**Question 1: What is the sum of the outputs of the generator for limit = 5?**

In [50]:
def square_root_generator(limit):
    n = 1
    while n <= limit:
        yield n ** 0.5
        n += 1

if __name__ == "__main__":
    limit = 5
    sum_sqrt = 0
    for sqrt_value in square_root_generator(limit):
        print(sqrt_value)
        sum_sqrt += sqrt_value
    print(sum_sqrt)

1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
8.382332347441762


Answer: 8.382332347441762

**Question 2: What is the 13th number yielded by the generator?**

In [52]:
def square_root_generator(limit):
    n = 1
    while n <= limit:
        yield n ** 0.5
        n += 1

limit = 13
generator = square_root_generator(limit)

for sqrt_value in generator:
    print(sqrt_value)

1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
2.449489742783178
2.6457513110645907
2.8284271247461903
3.0
3.1622776601683795
3.3166247903554
3.4641016151377544
3.605551275463989


Answer: 3.605551275463989

**Question 3: Append the 2 generators. After correctly appending the data, calculate the sum of all ages of people.**

In [53]:
def people_1():
    for i in range(1, 6):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 25 + i, "City": "City_A"}

for person in people_1():
    print(person)


def people_2():
    for i in range(3, 9):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 30 + i, "City": "City_B", "Occupation": f"Job_{i}"}


for person in people_2():
    print(person)

{'ID': 1, 'Name': 'Person_1', 'Age': 26, 'City': 'City_A'}
{'ID': 2, 'Name': 'Person_2', 'Age': 27, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 28, 'City': 'City_A'}
{'ID': 4, 'Name': 'Person_4', 'Age': 29, 'City': 'City_A'}
{'ID': 5, 'Name': 'Person_5', 'Age': 30, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 33, 'City': 'City_B', 'Occupation': 'Job_3'}
{'ID': 4, 'Name': 'Person_4', 'Age': 34, 'City': 'City_B', 'Occupation': 'Job_4'}
{'ID': 5, 'Name': 'Person_5', 'Age': 35, 'City': 'City_B', 'Occupation': 'Job_5'}
{'ID': 6, 'Name': 'Person_6', 'Age': 36, 'City': 'City_B', 'Occupation': 'Job_6'}
{'ID': 7, 'Name': 'Person_7', 'Age': 37, 'City': 'City_B', 'Occupation': 'Job_7'}
{'ID': 8, 'Name': 'Person_8', 'Age': 38, 'City': 'City_B', 'Occupation': 'Job_8'}


In [54]:
import dlt

pipeline = dlt.pipeline(destination='duckdb', dataset_name='all_people')

sum_age  = pipeline.run(people_1(),
										table_name="people_info",
										write_disposition="replace")

print(sum_age)

Pipeline dlt_colab_kernel_launcher load step completed in 0.22 seconds
1 load package(s) were loaded to destination duckdb and into dataset all_people
The duckdb destination used duckdb:////content/dlt_colab_kernel_launcher.duckdb location to store data
Load package 1707713086.084598 is LOADED and contains no failed jobs


In [55]:
import duckdb

conn = duckdb.connect(f"{pipeline.pipeline_name}.duckdb")

# let's see the tables
conn.sql(f"SET search_path = '{pipeline.dataset_name}'")
print('Loaded tables: ')
display(conn.sql("show tables"))

# and the data

print("\n\n\n people_info table below:")

people_info = conn.sql("SELECT * FROM people_info").df()
display(people_info)

Loaded tables: 


┌─────────────────────┐
│        name         │
│       varchar       │
├─────────────────────┤
│ _dlt_loads          │
│ _dlt_pipeline_state │
│ _dlt_version        │
│ people_info         │
└─────────────────────┘




 people_info table below:


id      name  age    city       _dlt_load_id         _dlt_id occupation
0   1  Person_1   26  City_A  1707713086.084598  PefpGj0npD4EHA       None
1   2  Person_2   27  City_A  1707713086.084598  MjYc5kUW6fCG+A       None
2   3  Person_3   28  City_A  1707713086.084598  nBcgYmiWQHBLwg       None
3   4  Person_4   29  City_A  1707713086.084598  VsaVsGzN2Xv5cw       None
4   5  Person_5   30  City_A  1707713086.084598  fBlpeDP6Q5GwQg       None

In [56]:
query_1 = "SELECT SUM(age) FROM people_info"
ages_sum = conn.sql(query_1).df()

display(ages_sum)

sum(age)
0     140.0

In [57]:
sum_age = pipeline.run(people_2(),
										table_name="people_info",
										write_disposition="append")

In [58]:
print(sum_age)

Pipeline dlt_colab_kernel_launcher load step completed in 0.22 seconds
1 load package(s) were loaded to destination duckdb and into dataset all_people
The duckdb destination used duckdb:////content/dlt_colab_kernel_launcher.duckdb location to store data
Load package 1707713094.962843 is LOADED and contains no failed jobs


In [59]:
import duckdb

conn = duckdb.connect(f"{pipeline.pipeline_name}.duckdb")

# let's see the tables
conn.sql(f"SET search_path = '{pipeline.dataset_name}'")
print('Loaded tables: ')
display(conn.sql("show tables"))

# and the data

print("\n\n\n people_info table below:")

people_info = conn.sql("SELECT * FROM people_info").df()
display(people_info)

Loaded tables: 


┌─────────────────────┐
│        name         │
│       varchar       │
├─────────────────────┤
│ _dlt_loads          │
│ _dlt_pipeline_state │
│ _dlt_version        │
│ people_info         │
└─────────────────────┘




 people_info table below:


id      name  age    city       _dlt_load_id         _dlt_id occupation
0    1  Person_1   26  City_A  1707713086.084598  PefpGj0npD4EHA       None
1    2  Person_2   27  City_A  1707713086.084598  MjYc5kUW6fCG+A       None
2    3  Person_3   28  City_A  1707713086.084598  nBcgYmiWQHBLwg       None
3    4  Person_4   29  City_A  1707713086.084598  VsaVsGzN2Xv5cw       None
4    5  Person_5   30  City_A  1707713086.084598  fBlpeDP6Q5GwQg       None
5    3  Person_3   33  City_B  1707713094.962843  20Ui7Rgnmidn/A      Job_3
6    4  Person_4   34  City_B  1707713094.962843  4w43Al47JE79KQ      Job_4
7    5  Person_5   35  City_B  1707713094.962843  33sJRnOPEnFm0w      Job_5
8    6  Person_6   36  City_B  1707713094.962843  niCbCva/Pnc9ig      Job_6
9    7  Person_7   37  City_B  1707713094.962843  nhSCuKn4u4Ew7w      Job_7
10   8  Person_8   38  City_B  1707713094.962843  yOOZok+nwGFZVw      Job_8

In [60]:
query_2 = "SELECT SUM(age) FROM people_info"
ages_sum = conn.sql(query_2).df()

display(ages_sum)

sum(age)
0     353.0

Answer: 353

**Question 4: Merge the 2 generators using the ID column. Calculate the sum of ages of all the people loaded as described above.**

In [61]:
def people_1():
    for i in range(1, 6):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 25 + i, "City": "City_A"}

for person in people_1():
    print(person)

def people_2():
    for i in range(3, 9):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 30 + i, "City": "City_B", "Occupation": f"Job_{i}"}

for person in people_2():
    print(person)

{'ID': 1, 'Name': 'Person_1', 'Age': 26, 'City': 'City_A'}
{'ID': 2, 'Name': 'Person_2', 'Age': 27, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 28, 'City': 'City_A'}
{'ID': 4, 'Name': 'Person_4', 'Age': 29, 'City': 'City_A'}
{'ID': 5, 'Name': 'Person_5', 'Age': 30, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 33, 'City': 'City_B', 'Occupation': 'Job_3'}
{'ID': 4, 'Name': 'Person_4', 'Age': 34, 'City': 'City_B', 'Occupation': 'Job_4'}
{'ID': 5, 'Name': 'Person_5', 'Age': 35, 'City': 'City_B', 'Occupation': 'Job_5'}
{'ID': 6, 'Name': 'Person_6', 'Age': 36, 'City': 'City_B', 'Occupation': 'Job_6'}
{'ID': 7, 'Name': 'Person_7', 'Age': 37, 'City': 'City_B', 'Occupation': 'Job_7'}
{'ID': 8, 'Name': 'Person_8', 'Age': 38, 'City': 'City_B', 'Occupation': 'Job_8'}


In [62]:
import dlt

pipeline = dlt.pipeline(destination='duckdb', dataset_name='all_people')

sum_age = pipeline.run(people_1(),
										table_name="people_info",
										write_disposition="replace")

sum_age = pipeline.run(people_2(),
										table_name="people_info",
										write_disposition="merge",
                    primary_key='ID')

print(sum_age)

Pipeline dlt_colab_kernel_launcher load step completed in 0.32 seconds
1 load package(s) were loaded to destination duckdb and into dataset all_people
The duckdb destination used duckdb:////content/dlt_colab_kernel_launcher.duckdb location to store data
Load package 1707713115.0956554 is LOADED and contains no failed jobs


In [63]:
import duckdb

conn = duckdb.connect(f"{pipeline.pipeline_name}.duckdb")

# let's see the tables
conn.sql(f"SET search_path = '{pipeline.dataset_name}'")
print('Loaded tables: ')
display(conn.sql("show tables"))

# and the data

print("\n\n\n people_info table below:")

people_info = conn.sql("SELECT * FROM people_info").df()
display(people_info)

Loaded tables: 


┌─────────────────────┐
│        name         │
│       varchar       │
├─────────────────────┤
│ _dlt_loads          │
│ _dlt_pipeline_state │
│ _dlt_version        │
│ people_info         │
└─────────────────────┘




 people_info table below:


id      name  age    city        _dlt_load_id         _dlt_id occupation
0   1  Person_1   26  City_A  1707713114.5145419  gyyQVqcQ6tCbug       None
1   2  Person_2   27  City_A  1707713114.5145419  te2Nzegd6pUhHQ       None
2   5  Person_5   35  City_B  1707713115.0956554  fpHj6exfeVREOQ      Job_5
3   8  Person_8   38  City_B  1707713115.0956554  zbNgK3imSxGMzQ      Job_8
4   7  Person_7   37  City_B  1707713115.0956554  Iv1g3jvoiJ9Pfg      Job_7
5   4  Person_4   34  City_B  1707713115.0956554  uPrhELLAF0a13w      Job_4
6   3  Person_3   33  City_B  1707713115.0956554  vhpVCEKJwwNJnw      Job_3
7   6  Person_6   36  City_B  1707713115.0956554  uUoyqhDgwiSEnw      Job_6

In [64]:
query_3 = "SELECT SUM(age) FROM people_info"
ages_sum = conn.sql(query_3).df()

display(ages_sum)

sum(age)
0     266.0

Answer: 266